doc/ipmc/IPMC_Data_Plane_HLD.md

```diff
@@ -0,0 +1,472 @@
+5. Support warm reboot for IP multicast routes
+
+#### Configuration and Management Requirements
+
+This document proposes that configuration and management should meet the following:
+
+1. Enable IP multicast on interfaces
+
+   > The default interface is not enabled to forward multicast routes
+
+### Architecture Design
+
+This document is designed to integrate the IP multicast routing function into the SWSS Container, so that related components can support IP multicast routing installing and warm reboot functions, and finally support the IP multicast routing function at the chip level.
+
+In SONiC, unicast routes and ECMP routes are sent via the fpm component to the APPL DB for further processing on the data plane. However, for IP multicast routing, the FRR Container has not passed the IP multicast routing data to the fpm component.
+
+The frr-pimd/pim6d daemon process is introduced in the FRR Container, in order to learn IP multicast routes under the PIM protocol and install IP multicast routes to the kernel.
+
+For the implementation of the data plane, firstly, the fpm component does not support IP multicast routing right now; secondly, in order to support more IP multicast protocols as much as possible;  thirdly, the implementation and support of FRR Container are beyond the design scope of this document, so the design will use Linux kernel as multicast route source.
+
+In ECMP routing, there is already a very elegant nexthop group design and implementation, so for IP multicast routing reference this design uses mgmanager to manage the ipmc group and rpf group.
+
+The following diagram summarizes the key structure of IPMC functionality in SONiC:
+
+![IPMC main structure1](images/ipmc_components.png "Figure 1: IPMC main structure")
+
+Figure 1. IPMC main structure.
+
+### High-Level Design
+
+#### Read IP multicast routes from the kernel
+
+When obtaining kernel IP multicast routing information through NetLink, you need to listen to the corresponding message type:
+
+Listen for NetLink message group type
+
+```text
+RTNLGRP_IPV4_MROUTE
+RTNLGRP_IPV6_MROUTE
+```
+
+Listens for the netlink message type
+
+```text
+RTM_NEWROUTE
+RTM_DELROUTE
+```
+
+The obtained routing information family should be of multicast type
+
+```text
+RTNL_FAMILY_IPMR
+```
+
+#### DB Changes
+
+##### APPL_DB changes
+
+Special note:
+
+1. The multicast route has one more key field than other routes: the source IP address. Therefore, the multicast route cannot share the same TABLE with the original routing entries
+2. When delivering an ipv6 multicast route, assume that both the source and destination IP addresses are ipv6. If the default delimiter of APPL DB ':' is used, the two ipv6 addresses cannot be distinguished by delimiter. Therefore, the delimiter '|' is used here.
+
+MROUTE_TABLE
+
+Producer: mroutesyncd
+
+Consumer: mrouteorch
+
+Description: A table was added to store IP multicast routing data. One entry corresponds to one multicast route.
+
+Schema:
+

```

Comment:
APP_DB schema is missing for INTF_TABLE to store multicast forwarding on interface.

Reply:
> APP_DB schema is missing for INTF_TABLE to store multicast forwarding on interface.

Yes, it will be revised in the next commit.